File: Data Structures & Algorithms/subsets/submission-0.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        dfs(0, nums, res, path);
        return res;
    }

    private void dfs(int i, int[] nums, List<List<Integer>> res, List<Integer> path) {
        // base case
        if (i == nums.length) {
            res.add(new ArrayList<>(path));
            return;
        }
        // 不选 nums[i]
        dfs(i + 1, nums, res, path);

        // 选nums[i]
        path.add(nums[i]);
        dfs(i + 1, nums, res, path);
        path.removeLast();
    }
}

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-0.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # (-1, n-1)
        l = -1
        r = len(nums) - 1
        
        while l + 1 < r:
            mid = (l + r) // 2
            if nums[mid] < nums[-1]:
                r = mid
            else:
                l = mid
        return nums[l + 1] # or nums[r]